bindings/c/include/vw/experimental/workspace.h

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include "vw/experimental/types.h"
+#include "vw/experimental/vwvis.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+  VW_DLL_PUBLIC VWStatus vw_create_workspace(VWOptions* options_handle, VWTraceMessageFunc* trace_listener,
+      void* trace_context, VWWorkspace** output_handle, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+  VW_DLL_PUBLIC VWStatus vw_create_workspace_with_model(VWOptions* options_handle, void* context, VWReadFunc* reader,
+      VWTraceMessageFunc* trace_listener, void* trace_context, VWWorkspace** output_handle,
+      VWErrorString* err_str_container) VW_API_NOEXCEPT;
+  VW_DLL_PUBLIC VWStatus vw_create_seeded_workspace(const VWWorkspace* existing_workspace_handle,
+      VWOptions* extra_options_handle, VWTraceMessageFunc* trace_listener, void* trace_context,
+      VWWorkspace** output_handle, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+
+  VW_DLL_PUBLIC VWStatus vw_workspace_get_model_id(
+      const VWWorkspace* workspace_handle, const char** model_id, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+  VW_DLL_PUBLIC VWStatus vw_workspace_set_model_id(
+      VWWorkspace* workspace_handle, const char* model_id, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+  VW_DLL_PUBLIC VWStatus vw_workspace_get_command_line(
+      const VWWorkspace* workspace_handle, VWString* command_line, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+
+  // finish is now broken up into finish() and destroy_workspace().
+  VW_DLL_PUBLIC VWStatus vw_workspace_finish(
+      VWWorkspace* workspace_handle, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+  VW_DLL_PUBLIC VWStatus vw_destroy_workspace(
+      VWWorkspace* workspace_handle, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+
+  VW_DLL_PUBLIC VWStatus vw_workspace_get_prediction_type(const VWWorkspace* workspace_handle,
+      VWPredictionType* prediction_type, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+  VW_DLL_PUBLIC VWStatus vw_workspace_get_label_type(
+      const VWWorkspace* workspace_handle, VWLabelType* label_type, VWErrorString* err_str_container) VW_API_NOEXCEPT;
+
+  // TODO add multipredict
+
+  // These are "Legacy" variants because predictions and labels are in the example object
```

Comment:
are these post-fixed with `legacy` to indicate that they will be changed in the near future?

Reply:
Yes, the goal is to introduce variants where label and prediction exist externally to the example object and as separate parameters. I used legacy so we don't have to break the API to move to that other interface. "legacy" is a bit icky though...

Reply:
I was also wondering about the fact that we use `workspace` everywhere here and if it is adding any value. Can there be a different `vw_<x>_learn_multiline` or something like that, if we pass in a different handle that is not `VWWorkspace`?

Reply:
I think we want the function names to be verbose. They are all operating as members of the workspace object. If a different object offered the ability to perform the learn operation it would offer a similar function